src/codegen/spirv.zig

```diff
@@ -988,11 +995,49 @@ const NavGen = struct {
-                            return self.todo("packed struct constants", .{});
+                        if (zcu.typeToPackedStruct(ty)) |struct_type| {
+                            comptime assert(Type.packed_struct_layout_version == 2);
+                            const backing_int_ty = Type.fromInterned(struct_type.backingIntTypeUnordered(ip));
+                            var running_int_id = try self.constInt(backing_int_ty, 0);
+                            var running_bits: u16 = 0;
+                            for (0..struct_type.field_types.len) |field_index| {
+                                const field_ty = Type.fromInterned(struct_type.field_types.get(ip)[field_index]);
+                                if (!field_ty.hasRuntimeBitsIgnoreComptime(zcu)) continue;
+                                const field_val = switch (ip.indexToKey(val.toIntern()).aggregate.storage) {
+                                    .bytes => |bytes| try pt.intern(.{ .int = .{
+                                        .ty = field_ty.toIntern(),
+                                        .storage = .{ .u64 = bytes.at(field_index, ip) },
+                                    } }),
+                                    .elems => |elems| elems[field_index],
+                                    .repeated_elem => |elem| elem,
+                                };
+                                const field_id = try self.constant(field_ty, Value.fromInterned(field_val), .direct);
+                                const ty_bit_size: u16 = @intCast(field_ty.bitSize(zcu));
+                                const field_int_ty = try self.pt.intType(.unsigned, ty_bit_size);
+                                const field_int_val = try self.bitCast(field_int_ty, field_ty, field_id);
+                                const shift_rhs = try self.constInt(backing_int_ty, running_bits);
+                                const extended_int_conv = try self.buildIntConvert(backing_int_ty, .{
+                                    .ty = field_int_ty,
+                                    .value = .{ .singleton = field_int_val },
+                                });
+                                const shifted = try self.buildBinary(.sll, extended_int_conv, .{
+                                    .ty = backing_int_ty,
+                                    .value = .{ .singleton = shift_rhs },
+                                });
+                                const running_int_tmp = try self.buildBinary(
+                                    .bit_or,
+                                    .{ .ty = backing_int_ty, .value = .{ .singleton = running_int_id } },
+                                    shifted,
+                                );
+                                running_int_id = try running_int_tmp.materialize(self);
+                                running_bits += ty_bit_size;
+                            }
+                            return running_int_id;
```

Comment:
Why is this constructing the int value at runtime rather than at compile time? Shouldn't there be a function which turns a packed struct constant into an integer, which you can then just lower into SPIR-V by constInt? Like what is currently already being done by the packed union thing?